File: server/src/main/java/gg/mineral/server/world/schematic/SchematicFile.java
package gg.mineral.server.world.schematic;

import java.io.File;
import java.util.Collections;
import java.util.List;

import gg.mineral.server.world.chunk.ChunkImpl;
import it.unimi.dsi.fastutil.shorts.Short2ObjectOpenHashMap;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

@Getter
@AllArgsConstructor
public class SchematicFile {
    private final Short2ObjectOpenHashMap<List<SchematicBlock>> chunkedBlocks = new Short2ObjectOpenHashMap<>();
    private final File source;
    @Setter
    private short xSize, ySize, zSize;

    public List<SchematicBlock> getBlocksForChunk(byte chunkX, byte chunkZ) {
        short key = ChunkImpl.toKey(chunkX, chunkZ);
        return chunkedBlocks.getOrDefault(key, Collections.emptyList());
    }
}

File: server/src/main/java/gg/mineral/server/world/schematic/SchematicFile.kt
package gg.mineral.server.world.schematic

import gg.mineral.server.world.chunk.ChunkImpl.Companion.toKey
import it.unimi.dsi.fastutil.shorts.Short2ObjectOpenHashMap
import lombok.AllArgsConstructor
import lombok.Getter
import lombok.Setter
import java.io.File

@Getter
@AllArgsConstructor
class SchematicFile {
    val chunkedBlocks: Short2ObjectOpenHashMap<List<SchematicBlock>> =
        object : Short2ObjectOpenHashMap<List<SchematicBlock?>?>() {
            init {
                defaultReturnValue(emptyList<SchematicBlock>())
            }
        }
    private val source: File? = null

    @Setter
    private val xSize: Short = 0

    @Setter
    private val ySize: Short = 0

    @Setter
    private val zSize: Short = 0

    fun getBlocksForChunk(chunkX: Byte, chunkZ: Byte): List<SchematicBlock> {
        val key = toKey(chunkX, chunkZ)
        return chunkedBlocks.remove(key)
    }

    val isEmpty: Boolean
        get() = chunkedBlocks.isEmpty()
}